zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1221,21 +1234,84 @@ private void reconnect() {
     }
   }
 
+
+
+  private void doAsyncSync(final ZooKeeper zk, final String path, final long startT,
+      final ZkAsyncCallbacks.SyncCallbackHandler cb) {
+    zk.sync(path, cb,
+        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
+          @Override
+          protected void doRetry() throws Exception {
+            doAsyncSync(zk, path, System.currentTimeMillis(), cb);
+          }
+        });
+  }
+
+  /*
+   *  Note, issueSync takes a ZooKeeper (client) object and pass it to doAsyncSync().
+   *  The reason we do this is that we want to ensure each new session event is preceded with exactly
+   *  one sync() to server. The sync() is to make sure the server would not see stale data.
+   *
+   *  ZooKeeper client object has an invariant of each object has one session. With this invariant
+   *  we can achieve each one sync() to server upon new session establishment. The reasoning is:
+   *  issueSync() is called when fireNewSessionEvents() which in under eventLock of ZkClient. Thus
+   *  we are guaranteed the ZooKeeper object passed in would have the new incoming sessionId. If by
+   *  the time sync() is invoked, the session expires. The sync() would fail with a stale session.
+   *  This is exactly what we want. The newer session would ensure another fireNewSessionEvents.
+   */
+  private boolean issueSync(String sessionId, ZooKeeper zk) throws ZkInterruptedException {
```

Comment:
I still think Zookeeper is not required here. Please see my reply to the original thread. If we put the check logic here, it will work.

The reason I insist on doing this is that, when you put the old zk object here, if any new session created, then the old zk object will throw exceptions when you call it. That would be a ZK exception. And then we rely on our exception handling logic to ensure everything works fine.

If we check the current zk object with the session Id, then we can end this operation with explicit logic and with a very clean warning or error log.
Moreover, as I mentioned in the first comment, the session Id and the input zk contain redundant information. You only need one of them.

Reply:
See my reply (just now) of the after fact checking cons here: 

> Just saw this comment.
The gist is that the checking of sessionId and doAsyncSync() is not atomic. After if (zk.getSessionId().equals(sessionId)) succeed, the Zookeeper object may be changed due to a new session established. This would make the current doAsyncSync() work for the new session. Since the new session would trigger another fireNewSession(), which would put another retrySync(). Then we will have another sync() to the Zookeeper for the new session. So in sum, the new session can be two sync() in a row.

Back to your concern about old Zk object throw exceptions, here is how I look at it:

First, async call (sync() in this case), would not throw exception. The error code would from the callback (invoked in eventthread of ZooKeeper) would tell what is wrong. The checking is in `protected boolean needRetry(int rc)`.

Reply:
“Please see my reply to the original thread.”, where I mentioned that this con is not real : )

When you get the zk first then compared it with the session Id, you will have the right zk to use for sure. Or the check will fail and we will not really send sync request.

Reply:
As synced offline, we will not pass redundant data to issueAsync(). Only add zk.